pkg/image/layer_metadata.go

```diff
@@ -16,17 +16,19 @@ type LayerMetadata struct {
 }
 
 // newLayerMetadata aggregates pertinent layer metadata information.
-func newLayerMetadata(imgMetadata Metadata, layer v1.Layer, idx int) (LayerMetadata, error) {
+func newLayerMetadata(layer v1.Layer, idx int) (LayerMetadata, error) {
 	mediaType, err := layer.MediaType()
 	if err != nil {
 		return LayerMetadata{}, err
 	}
+	diffID, err := layer.DiffID()
+	if err != nil {
+		return LayerMetadata{}, err
+	}
 
-	// digest = diff-id = a digest of the uncompressed layer content
-	diffIDHash := imgMetadata.Config.RootFS.DiffIDs[idx]
 	return LayerMetadata{
 		Index:     uint(idx),
-		Digest:    diffIDHash.String(),
+		Digest:    diffID.String(),
```

Comment:
I believe the most important part of this change is here, where there was a panic on line 26. Are we sure that in all cases the `layer.DiffID()` is going to return the same value (e.g. docker-archive, oci-archive, oci-directory, etc.)? I really don't know the right answer, but would be really wary about changing this logic. Could this PR just be simplified a bit to maybe check the `imgMetadata` and sub-properties and return an error in the event they are nil? Or _then_ try the `layer.DiffID`?

Reply:
Thanks for the feedback here @kzantow 

1. Do you know why stereoscope's code is reading the diff ID from the image metadata, rather than asking the layer for the diff ID? The panic seems like evidence that we're handling this incorrectly.
2. If indexing reading into the rootFS diff IDs returns a different answer than layer.DiffID, isn't the layer right?

I can revert this, and maybe I should. The panic will actually be fixed in this PR by checking the media type before calling `newLayerMetadata`.

The panic was because there is no RootFS for helm charts packed as OCI images, AFAICT. Calling layer.DiffID() on one returns an error.

Reply:
@kzantow So I did the following experiment:

In my local code, I added a function like this:

``` go
func panicOnDiffIDMismatch(l *Layer, imgMetadata Metadata, idx int) {
	interfaceDiffID, err := l.layer.DiffID()
	if err != nil {
		panic(err)
	}

	homebrewDiffID := imgMetadata.Config.RootFS.DiffIDs[idx]

	if homebrewDiffID != interfaceDiffID {
		panic("everything we know is wrong!")
	}
}
```

and called it at the top of `layer.Read`.

With this change included, Syft and Stereoscope integration tests pass, as long as I set `CONTAINER_HOST` to the socket podman is actually listening on, which is a requirement on `main` as well AFAICT.

I think I'm in favor of changing our implementation to use `layer.DiffID()` instead of our own diff ID:

1. If the two methods are always the same, it doesn't matter
2. If the two methods _aren't_ always the same, isn't the one implemented in `go-containerregistry` more likely correct?

On the other hand, I understand your caution about changing this. What do you think?

Reply:
agreed to use `layer.DiffID`, the original reasons for not doing so are lost to history (though it might have been for performance, but I'm not seeing anything in a code review in GGCR for this to be true).